crates/solver/src/settlement.rs

```diff
@@ -12,11 +13,23 @@ use primitive_types::{H160, U256};
 use shared::conversions::U256Ext as _;
 use std::collections::HashMap;
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum TokenIndex {
+    UniformClearingPriceIndex(usize),
+    CustomBuyPriceIndex(usize),
+}
```

Comment:
Probably too much for this PR but I feel like settlement.rs and settlement_encoder.rs are in need of a refactor. Having to special case it like this feels weird. Maybe we could keep track of liquidity orders separately so they can use their own struct, vector, prices, etc.

Reply:
Yeah..., agree.

What do you think about the following approach:
```

enum Trade{
Liquidity_Trade(TradeInner),
NormalTrade(TradeInner),
}

#[derive(Debug, Clone)]
pub struct SettlementEncoder {
    tokens: Vec<H160>,
    clearing_prices: HashMap<H160, U256>,
    trades: Vec<Trade>,
    execution_plan: Vec<Arc<dyn Interaction>>,
    unwraps: Vec<UnwrapWethInteraction>,
}

```

and then build the final clearing prices for liquidity orders only at the very end within the `finish` function.

That would be simplifying it, as then we don't need to store the non_uniform_clearing prices  in the settlementEncoder

Reply:
I could see this working. Although we have some code that currently iterates over all orders and does something with them. That would be harder. One alternative I meant was to not have enum. So there would be
Vec<NormalTrade>
Vec<LiquidityTrade>
and separate functions to get the from the encoder. This is hard to say without trying it out though.

Reply:
I also think we can refactor the settlement and settlement encoder to make this use-case fit in as I think the uniform clearing prices were somewhat baked into the assumptions of this module.

Maybe separate `Vec`s makes it easier here? I'm also not sure.

Reply:
```
Vec<NormalTrade>
Vec<LiquidityTrade>
```
I went for this suggestion, just with a little bit longer names:
```
Vec<OrderTrade>
Vec<LiquidityOrderTrade>
```
It seems to work nice.

I was also trying out another approach:
Stripping away the `buy_token_index` and `sell_token_index` in the `struct Trade` and only finding the indexes during the finish() function of the `SettlementEncoder`. This worked nicely, as then the `Trade` struct of  a liquidity order and a normal order do not differ. Then also, the work of maintaining the indices would not be needed during merges etc. 
But it had a major drawback: If the `SettlementEncoder` would not be valid - as some token indices do not exist - it would only fail during the finish() functionality. In the end, I found it cleaner to not have these hidden assumptions and go with your proposal.